### PR TITLE
Assert against chan.listen() on sub-channels

### PR DESCRIPTION
**Pull request by benfleis**
causes unexpected test failure, needs tweaks.

test failure here:

  https://github.com/uber/tchannel-node/issues/30#issuecomment-144597268

### CODE REVIEW

File: channel.js [hunk 1]
```diff
@@ -559,6 +559,7 @@ TChannel.prototype.listen = function listen(port, host, callback) {
     assert(typeof host === 'string', 'TChannel requires host argument');
     assert(typeof port === 'number', 'TChannel must listen with numeric port');
     assert(host !== '0.0.0.0', 'TChannel must listen with externally visible host');
+    assert(!self.topChannel, 'TChannel must listen on top channel');
```
Comment (jcorbin) [inline channel.js]: This is the opposite assertion of what you want:
- chan.topChannel is not a bool
- chan.topChannel is a reference to the top channel, if chan is a sub channel

**Comment by jcorbin** [comment]
Rebased and corrected your check, but yes this is a more useful form of "fail early" that we should have.

File: channel.js (hunk 1 above)
Comment (jcorbin) [inline channel.js]: Oops, I was exactly wrong... you had it right the first time.

**Comment by jcorbin** [comment]
I'll take this over and complete it from here; what you missed on the test failure was that you made the benchmark listen correctly on the root channel, but then the quit/close was still happening on the sub channel.

Diving through this code path highlight other room for hardening that I'll pull out first, and an open question about do we want a similar assertion for channel close.

cc @Raynos @kriskowal

**Comment by Raynos** [comment]
lgtm.

**Comment by Raynos** [comment]
:+1: for `_close()` and assertions in `close()`

**Comment by jcorbin** [comment]
TODO -> TODONE by splitting close and _close, will land once tests pass

**Comment by benfleis** [comment]
awesome, glad to see this picked up :)

**Comment by jcorbin** [comment]
TODO: this breaks hyperbahn integration tests, we need to update hyperbahn

**Comment by jcorbin** [comment]
Note: to dev iterate on this:
```
$ ./hyperbahn_link_test.sh ../hyperbahn # assuming that hyperbahn is checked out at ../hyperbahn
...
assert.js:92
  throw new assert.AssertionError({
  ^
AssertionError: cannot close sub-TChannel
    at TChannel.close (/Users/joshua/Uber/tchannel-node/channel.js:852:5)
    at ServiceDispatchHandler.purgeServices (/Users/joshua/Uber/hyperbahn/service-proxy.js:9:23057)
    at purgeServices [as _onTimeout] (/Users/joshua/Uber/hyperbahn/service-proxy.js:9:23637)
    at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
...
```

**Comment by Raynos** [comment]
Hmm; we close sub channels when a serviceName is gone. 

We should split this PR into two PRs; one making listen() invalid on sub channel and the other making close() invalid.

**Pull request merged**